File: claude-code-python/src/claude_code/constants/files.py
"""
File extension & binary detection utilities
原始 TS: src/constants/files.ts
"""
from __future__ import annotations

import os
from pathlib import Path

# ---------------------------------------------------------------------------
# Binary file extensions to skip for text-based operations
# ---------------------------------------------------------------------------

BINARY_EXTENSIONS: frozenset[str] = frozenset({
    # Images
    ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".ico",
    ".webp", ".tiff", ".tif",
    # Videos
    ".mp4", ".mov", ".avi", ".mkv", ".webm", ".wmv",
    ".flv", ".m4v", ".mpeg", ".mpg",
    # Audio
    ".mp3", ".wav", ".ogg", ".flac", ".aac",
    ".m4a", ".wma", ".aiff", ".opus",
    # Archives
    ".zip", ".tar", ".gz", ".bz2", ".7z", ".rar",
    ".xz", ".z", ".tgz", ".iso",
    # Executables/binaries
    ".exe", ".dll", ".so", ".dylib", ".bin",
    ".o", ".a", ".obj", ".lib", ".app",
    ".msi", ".deb", ".rpm",
    # Documents (PDF is here; FileReadTool excludes it at the call site)
    ".pdf", ".doc", ".docx", ".xls", ".xlsx",
    ".ppt", ".pptx", ".odt", ".ods", ".odp",
    # Fonts
    ".ttf", ".otf", ".woff", ".woff2", ".eot",
    # Bytecode / VM artifacts
    ".pyc", ".pyo", ".class", ".jar", ".war", ".ear",
    ".node", ".wasm", ".rlib",
    # Database files
    ".sqlite", ".sqlite3", ".db", ".mdb", ".idx",
    # Design / 3D
    ".psd", ".ai", ".eps", ".sketch", ".fig",
    ".xd", ".blend", ".3ds", ".max",
    # Flash
    ".swf", ".fla",
    # Lock/profiling data
    ".lockb", ".dat", ".data",
})

_BINARY_CHECK_SIZE = 8192


def has_binary_extension(file_path: str) -> bool:
    """Check if a file path has a binary extension."""
    ext = Path(file_path).suffix.lower()
    return ext in BINARY_EXTENSIONS


def is_binary_content(data: bytes) -> bool:
    """
    Check if a buffer contains binary content by looking for null bytes
    or a high proportion of non-printable characters.
    """
    check_size = min(len(data), _BINARY_CHECK_SIZE)
    if check_size == 0:
        return False

    non_printable = 0
    for i in range(check_size):
        byte = data[i]
        # Null byte is a strong indicator of binary
        if byte == 0:
            return True
        # Count non-printable, non-whitespace bytes
        # Printable ASCII: 32-126, plus common whitespace (9=tab, 10=LF, 13=CR)
        if byte < 32 and byte not in (9, 10, 13):
            non_printable += 1

    return (non_printable / check_size) > 0.1

File: claude-code-rust/crates/claude-constants/src/files.rs
// Original TS source: constants/files.ts
// File extension and binary content detection

use std::collections::HashSet;
use once_cell::sync::Lazy;

/// Set of binary file extensions to skip for text-based operations.
pub static BINARY_EXTENSIONS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    [
        // Images
        ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".ico", ".webp", ".tiff", ".tif",
        // Videos
        ".mp4", ".mov", ".avi", ".mkv", ".webm", ".wmv", ".flv", ".m4v", ".mpeg", ".mpg",
        // Audio
        ".mp3", ".wav", ".ogg", ".flac", ".aac", ".m4a", ".wma", ".aiff", ".opus",
        // Archives
        ".zip", ".tar", ".gz", ".bz2", ".7z", ".rar", ".xz", ".z", ".tgz", ".iso",
        // Executables/binaries
        ".exe", ".dll", ".so", ".dylib", ".bin", ".o", ".a", ".obj", ".lib", ".app",
        ".msi", ".deb", ".rpm",
        // Documents
        ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx", ".odt", ".ods", ".odp",
        // Fonts
        ".ttf", ".otf", ".woff", ".woff2", ".eot",
        // Bytecode / VM artifacts
        ".pyc", ".pyo", ".class", ".jar", ".war", ".ear", ".node", ".wasm", ".rlib",
        // Database files
        ".sqlite", ".sqlite3", ".db", ".mdb", ".idx",
        // Design / 3D
        ".psd", ".ai", ".eps", ".sketch", ".fig", ".xd", ".blend", ".3ds", ".max",
        // Flash
        ".swf", ".fla",
        // Lock/profiling data
        ".lockb", ".dat", ".data",
    ]
    .iter()
    .cloned()
    .collect()
});

/// Check if a file path has a binary extension.
pub fn has_binary_extension(file_path: &str) -> bool {
    if let Some(dot_pos) = file_path.rfind('.') {
        let ext = &file_path[dot_pos..].to_lowercase();
        BINARY_EXTENSIONS.contains(ext.as_str())
    } else {
        false
    }
}

const BINARY_CHECK_SIZE: usize = 8192;

/// Check if a buffer contains binary content by looking for null bytes
/// or a high proportion of non-printable characters.
pub fn is_binary_content(buffer: &[u8]) -> bool {
    let check_size = buffer.len().min(BINARY_CHECK_SIZE);
    let buf = &buffer[..check_size];

    let mut non_printable = 0usize;
    for &byte in buf {
        if byte == 0 {
            return true;
        }
        if byte < 32 && byte != 9 && byte != 10 && byte != 13 {
            non_printable += 1;
        }
    }

    if check_size == 0 {
        return false;
    }

    (non_printable as f64) / (check_size as f64) > 0.1
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_has_binary_extension() {
        assert!(has_binary_extension("image.png"));
        assert!(has_binary_extension("video.mp4"));
        assert!(!has_binary_extension("main.rs"));
        assert!(!has_binary_extension("readme.md"));
    }

    #[test]
    fn test_is_binary_content() {
        let text = b"Hello, world! This is plain text.\n";
        assert!(!is_binary_content(text));

        let binary = &[0u8, 1, 2, 3, 4];
        assert!(is_binary_content(binary));
    }
}
